### DISCUSSION TITLE
MQTT using Sparkplug B

### DISCUSSION THREAD

**Discussion by allnerag** [arendst/Tasmota#15351] [Ideas]
Good day,

Do we see the need to bring in support for "SparkPlug B" to be able to use any easy accessible sensor in the OT (OT network is industrial-oriented,) network with MQTT by using the extension.  I think we would then be able to integrate custom build sensors into any OT type of software like. We can do this now when we integrate with Node Red to convert from MQTT to MQTT with SparkPlug B. This is now an additional work to be done and would like if it was native.

https://www.hivemq.com/solutions/technology/mqtt-sparkplug/

Thanks

**Comment by barbudor** [comment]
A standard MQTT topic structure insuring interoperability could be interresting.
Is that achievable in Tasmota without breaking everything is a more complex question.

So far from th elinks you provided I can just see blahblah whitepapers and nothing concrete in terms of topic structure specification
Have you found anything more solid ? A real specification ?

**Reply by thomassorensen2650** [reply]
Sparkplub B support would enable out-of-the-box integration with a lot of systems and would really be a amazing feature.

The official specification for Sparkplug B can be found here: https://github.com/eclipse/sparkplug/tree/develop/specification/src/main/asciidoc/chapters

**Reply by sfromis** [reply]
New features are usually added to Tasmota as a result of someone interested sitting down to do the work of exploring and/or coding. You?

**Reply by motodd01** [reply]
I will look into this. Sparkplugb is really amazing. It is the backbone of the IIOT system I developed at my day job.

**Reply by michcfr** [reply]
> A standard MQTT topic structure insuring interoperability could be interresting. Is that achievable in Tasmota without breaking everything is a more complex question.
> 
> So far from th elinks you provided I can just see blahblah whitepapers and nothing concrete in terms of topic structure specification Have you found anything more solid ? A real specification ?

Hello,
Maybe should we reactivate the discussion?

Things has evolved with Sparkplug, Client libraries and reference implementations exists now:
-general for various languages and for various devices: https://github.com/eclipse/tahu
-under Arduino: https://github.com/mkeras/EmbeddedSparkplugNode

Any chance to adopt it under Tasmota?

**Reply by sfromis** [reply]
As already said:
> New features are usually added to Tasmota as a result of someone interested sitting down to do the work of exploring and/or coding. You?

Chances depends on who's "in the mood" for the task, and if it fits well with the structure of Tasmota.

**Reply by michcfr** [reply]
> As already said:
> 
> > New features are usually added to Tasmota as a result of someone interested sitting down to do the work of exploring and/or coding. You?
> 
> Chances depends on who's "in the mood" for the task, and if it fits well with the structure of Tasmota.

We don't have to push at open doors here, we are all agreed on that.

This a request for advices from Tasmota expert to start the discussion about the technical feasability and how this task can be done. For example, to start the discussion, use C code or Berry?

**Reply by barbudor** [reply]
Berry will target only ESP32 and will not replace Tasmota native MQTT scheme
Which is also a big question: Should Sparkplug comes in addition or as a replacement of Tasmota's native scheme ?
If it has to replace, it has to cover all main MQTT activity of Tasmota : cmnd, stat and tele

When I see the type of messages, I feel the match might not be easy to achieve

**Reply by michcfr** [reply]
ok so you mean it is not more than MQTT messages translation of the Tasmota native scheme to and from SparkPlug? it is rather a complete rewriting? 
Is the existing Tasmota MQTT code be separated into layers: one specific to Tasmota and the other specific to the MQTT output: native scheme or SparkPlug? At least the Tasmota MQTT can be modular so that we can choose (at compilation with #define?) between the native or the sparkplug?

**Reply by sfromis** [reply]
As the "M" in the name "Tasmota" literally means "MQTT", this is not isolated into a plug-able component, like when the data format JSON is handled throughout the relevant places in the Tasmota code, including the features allowing users to create commands, rules, scripter and Berry programs. While the MQTT protocol handling is only a few places, that does not go for the data format and control flows.

The question would be how to map between the Tasmota MQTT design and whatever message structure and flows SparkPlug requires. If this would not be straightforward to do in a separate bridge based on the existing MQTT communication, it would likely not be practical to integrate into Tasmota with a `#define` to switch which way. And the latter is my guess. But if a separate bridge would be feasible, that would be my suggestion.

Still, moving forward with this would need analysis by someone familiar with both sides, and with a strong interest in making it work with SparkPlug. And willing to do a significant (my guess) amount of work.

**Reply by michcfr** [reply]
In other words, it means it would heavily break the Tasmota code to move to SparkPlug?

**Reply by s-hadinger** [reply]
Not necessarily. After a quick look, it seems that all Sparkplug topics are in the form `spBv1.0/[Group ID]/[Message Type]/[EON Node ID]/[Device ID]` which does not interfere with Tasmota topics.

But it would require a whole new layer of Sparkplug interface, and largely doable in Berry.

It looks to me an overkill for home automation. Sparkplug is designed for SCADA and industrial usage. Again any volunteer is welcome.
